presto-cli/src/main/java/com/facebook/presto/cli/Console.java

```diff
@@ -356,6 +357,14 @@ private static boolean process(QueryRunner queryRunner, String sql, OutputFormat
                 builder = builder.withPreparedStatements(preparedStatements);
             }
 
+            // update session functions if present
+            if (!query.getAddedSessionFunctions().isEmpty() || !query.getRemovedSessionFunctions().isEmpty()) {
+                Map<String, String> sessionFunctions = new HashMap<>(session.getSessionFunctions());
+                sessionFunctions.putAll(query.getAddedSessionFunctions());
+                sessionFunctions.keySet().removeAll(query.getRemovedSessionFunctions());
```

Comment:
This might be masking ACL issues? Or is past that?

Reply:
This is just a parsed representation of the SQL function (no analysis at this point), but when a query containing this function is submitted for execution, the ACLs will be checked at analysis time.

Reply:
What would happen if I create `foo`, drop `foo` then create `foo` again? Why do we need to track removed functions rather than just removing them?

Reply:
Create foo, drop foo, create foo again would work just fine. The design of console is such that the server always sends back diffs to the client state rather than the full client state. For example, session properties and prepared statements also work this way.